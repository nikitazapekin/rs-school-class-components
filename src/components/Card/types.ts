 
export interface CardProps {

item: {

       name: string;
       email: string;
        password: string;
        age: number;
        gender: string;
        country: string;
        agreeToTerms: boolean;
        confirmPassword: string;
        avatar: string;
    }


  }