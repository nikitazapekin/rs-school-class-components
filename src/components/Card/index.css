.user {
    padding: 40px 20px;
    font-size: 32px;
    background-color: rgb(201, 174, 124);
    display: flex;
    transition: 0.5s ease-in-out;
    align-items: flex-start;
    cursor: pointer;
}

.user__list {
    display: flex;
    flex-direction: column;
    row-gap: 20px;
}

.user__logo {
    border-radius: 50%;
    max-width: 300px;
}

.user__link {
    text-decoration: none;
    color: #000;
}

.user-dark {
    background-color: rgb(73, 71, 71);
    color: #fff;
}

.user:hover {
    background-color: #92411b;
}

.user__content {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    row-gap: 20px;
    margin-left: 20px;
}

.user__login {
    font-size: 1.2rem;
    font-weight: bold;
}

.user__checkbox {
    position: relative;
    width: 20px;
    height: 20px;
    background-color: #fff;
    border: 1px solid #ccc;
    border-radius: 4px;
    cursor: pointer;
    outline: none;
    transition: background-color 0.3s;
}

.user__checkbox:hover {
    background-color: #f0f0f0;
}

.user__details {
    font-size: 24px;
    cursor: pointer;
    padding: 10px 25px;
    transition: 1s;
}

.user__details:hover {
    background-color: rgb(90, 90, 190);
    color: #fff;
}
