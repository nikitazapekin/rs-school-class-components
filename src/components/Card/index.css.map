{"version":3,"sources":["index.scss","index.css"],"names":[],"mappings":"AAAA;EACC,kBAAA;EACA,eAAA;EACA,mCAAA;EACA,aAAA;EACA,4BAAA;EACA,uBAAA;EACA,eAAA;ACCD;ADCC;EACC,aAAA;EACA,sBAAA;EACA,aAAA;ACCF;ADEC;EACC,kBAAA;EACA,gBAAA;ACAF;ADGC;EACC,qBAAA;EACA,WAAA;ACDF;;ADKA;EACC,yBAAA;ACFD","file":"index.css"}