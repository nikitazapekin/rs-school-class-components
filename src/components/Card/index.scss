.card {
    display: flex;
    background-color: rgb(167, 150, 120);
    flex-direction: column;
    row-gap: 20px;
    padding: 40px;
    border-radius: 20px;
    &__item {
        font-size: 22px;
    }
  
}