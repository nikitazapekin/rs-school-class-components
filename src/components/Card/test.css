.user {
    background-color: red;
    width: 500px;
    height: 400px;
}