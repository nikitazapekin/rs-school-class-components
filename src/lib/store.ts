//lib/store.ts

import { combineReducers, configureStore } from '@reduxjs/toolkit';
import type { ThunkAction, Action } from '@reduxjs/toolkit';
import appSlice from './slices/appSlice';
import { setupListeners } from '@reduxjs/toolkit/query';
import { githubApi } from './slices/querySlice';
import { usersGithubApi } from './slices/userQuerySlice';
import selectedElementsSlice from './slices/selectedElementsSlice';
import testSlice from './slices/testSlice';

const rootReducer = combineReducers({
  appSlice: appSlice,
  selectedElementsSlice: selectedElementsSlice,
  testSlice: testSlice,
  [githubApi.reducerPath]: githubApi.reducer,
  [usersGithubApi.reducerPath]: usersGithubApi.reducer,
});

export const makeStore = () => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(githubApi.middleware).concat(usersGithubApi.middleware),
  });
};

// Infer the type of makeStore
export type AppStore = ReturnType<typeof makeStore>;
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<AppStore['getState']>;
export type AppDispatch = AppStore['dispatch'];

// Set up listeners
const store = makeStore();
setupListeners(store.dispatch);

export default store;

/*import { configureStore } from '@reduxjs/toolkit'

export const makeStore = () => {
  return configureStore({
    reducer: {},
  })
}

// Infer the type of makeStore
export type AppStore = ReturnType<typeof makeStore>
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<AppStore['getState']>
export type AppDispatch = AppStore['dispatch']
*/




/*
import { combineReducers, configureStore } from '@reduxjs/toolkit';
import type { ThunkAction, Action } from '@reduxjs/toolkit';
import appSlice from './slices/appSlice';
import { setupListeners } from '@reduxjs/toolkit/query';
import { githubApi } from './slices/querySlice';
import { usersGithubApi } from './slices/userQuerySlice';
import selectedElementsSlice from './slices/selectedElementsSlice';

export const rootReducer = combineReducers({
	appSlice: appSlice,
	selectedElementsSlice: selectedElementsSlice,
	[githubApi.reducerPath]: githubApi.reducer,
	[usersGithubApi.reducerPath]: usersGithubApi.reducer,
});

export const store = configureStore({
	reducer: rootReducer,
	middleware: (getDefaultMiddleware) =>
		getDefaultMiddleware().concat(githubApi.middleware).concat(usersGithubApi.middleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;

setupListeners(store.dispatch);

*/