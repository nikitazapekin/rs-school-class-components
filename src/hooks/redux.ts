import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { RootState, AppDispatch } from '../redux/store';
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;


/*
import React from 'react';
import { render, renderHook } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { useAppDispatch, useAppSelector } from './redux'; // Убедитесь, что путь к файлу корректный
import { RootState, AppDispatch, rootReducer } from '../redux/store'; // Убедитесь, что путь к файлу корректный
import Providers from '../redux/Provider';
// Создание тестового Redux хранилища с правильным редьюсером
const testStore = configureStore({
  reducer: rootReducer,
  preloadedState: {
    appSlice: {
        isLoading: false,
	error: null,
	isLoadingUserData: false,
	users: [],
	typedValue: "",
	params: {
		limit: 10,
		offset: 1,
		query: '',
		storedValue: '',
	},
	status: 'idle',
	clickedUser: {
		login: "",
	id: 0,
	node_id:"" ,
	avatar_url: "",
	gravatar_id:"" ,
	url:"" ,
	html_url:"" ,
	followers_url: "",
	following_url:"" ,
	gists_url: "",
	starred_url: "",
	subscriptions_url:  "",
	organizations_url: "",
	repos_url:"" ,
	events_url:"" ,
	received_events_url:"" ,
	type: "",
	site_admin: false,
	score: 0
	}
   
       },
      selectedElementsSlice: {
        storedElements: [],

	selectedElement: null,
      },
      themeSlice: { 
        isDark: false
       }
  },
});

// Тестирование useAppDispatch
describe('useAppDispatch', () => {
  test('returns the correct dispatch type', () => {
    const { result } = renderHook(() => useAppDispatch(), {
      wrapper: ({ children }) => <Provider store={testStore}>{children}</Provider>,
    });

    expect(result.current).toBeDefined();
    expect(result.current).toEqual(expect.any(Function)); // Должен быть функцией
  });
});

// Тестирование useAppSelector
describe('useAppSelector', () => {
  test('returns the correct selector type', () => {
    // Определим начальное состояние для тестов
    const initialState: RootState = {
      appSlice: {
        isLoading: false,
	error: null,
	isLoadingUserData: false,
	users: [],
	typedValue: "",
	params: {
		limit: 10,
		offset: 1,
		query: '',
		storedValue: '',
	},
	status: 'idle',
	clickedUser: {
		login: "",
	id: 0,
	node_id:"" ,
	avatar_url: "",
	gravatar_id:"" ,
	url:"" ,
	html_url:"" ,
	followers_url: "",
	following_url:"" ,
	gists_url: "",
	starred_url: "",
	subscriptions_url:  "",
	organizations_url: "",
	repos_url:"" ,
	events_url:"" ,
	received_events_url:"" ,
	type: "",
	site_admin: false,
	score: 0
	}
   
       },
      selectedElementsSlice: {
        storedElements: [],

	selectedElement: null,
      },
      themeSlice: { 
        isDark: false
       }
    };

    // Обновим тестовый store с начальным состоянием
    const { result } = renderHook(() => useAppSelector((state: RootState) => state.appSlice), {
      wrapper: ({ children }) => <Provider store={testStore}>{children}</Provider>,
    });

    // Замените 'initialState.appSlice' на актуальные данные из начального состояния
    expect(result.current).toEqual(initialState.appSlice);
  });
});

*/