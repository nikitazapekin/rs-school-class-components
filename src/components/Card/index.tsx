import { useNavigate } from 'react-router-dom';
import './index.scss';

import ThemeContext from '../ThemeContext';
import { useContext } from 'react';
import { useAppDispatch } from '../../hooks/redux';
//import { useAppDispatch } from '@/hooks/redux.ts';
import { AddElementToStorage } from '../../lib/action-creators/addElementToStorage';
//import { AddElementToStorage } from '@/store/action-creators/addElementToStorage.ts';
import { useSelector } from 'react-redux';
//import { storedUsersSelector } from '@/store/selectors/getStoredElements';
import { storedUsersSelector } from '../../lib/selectors/getStoredElements';
import { UserElement, UserItem } from './types';
const Card = ({ user }: UserItem) => {
//	const navigate = useNavigate();
	const dispatch = useAppDispatch();
	const { isDark } = useContext(ThemeContext);
	const users = useSelector(storedUsersSelector);

	const handleCardClick = () => {
		const currentParams = new URLSearchParams(location.search);
		const currentUrl = `${window.location.pathname}?${currentParams.toString()}`;
		if (!currentUrl.includes('userdata')) {
			localStorage.setItem('lastUrl', currentUrl);
		}
		//navigate(`/main/userdata?username=${user.login}`);
	};

	const handleCheckboxChange = () => {
		dispatch(AddElementToStorage(user));
	};

	return (
		<div className="user__link">
			<div className={`user ${isDark ? `user-dark` : ``}`}>
				<img className="user__logo" src={user.avatar_url} alt="user" />
				<div className="user__content">
					<p className="user__login">{user.login}</p>
					<input
						type="checkbox"
						className="user__checkbox"
						onChange={handleCheckboxChange}
						checked={users && users.some((storedUser: UserElement) => user.id === storedUser.id)}
						readOnly
					/>
					<button className="user__details" onClick={handleCardClick}>
						Show details
					</button>
				</div>
			</div>
		</div>
	);
};

export default Card;
