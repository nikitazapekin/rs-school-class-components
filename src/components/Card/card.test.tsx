import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { rootReducer, RootState } from '../../redux/store';
import Card from './index';
import { createSerializer } from '@emotion/jest';  

expect.addSnapshotSerializer(createSerializer()); 

const initialState: RootState = {
  appSlice: {
    isLoading: false,
    error: null,
    isLoadingUserData: false,
    users: [],
    typedValue: "",
    params: {
      limit: 10,
      offset: 1,
      query: '',
      storedValue: '',
    },
    status: 'idle',
    clickedUser: {
      login: "",
      id: 0,
      node_id: "",
      avatar_url: "",
      gravatar_id: "",
      url: "",
      html_url: "",
      followers_url: "",
      following_url: "",
      gists_url: "",
      starred_url: "",
      subscriptions_url: "",
      organizations_url: "",
      repos_url: "",
      events_url: "",
      received_events_url: "",
      type: "",
      site_admin: false,
      score: 0
    }
  },
  selectedElementsSlice: {
    storedElements: [],
    selectedElement: null,
  },
  themeSlice: {
    isDark: false, 
  },
};

const renderWithRedux = (
  component: React.ReactNode,
  { store = configureStore({ reducer: rootReducer, preloadedState: initialState }) } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
};

describe('Card component', () => {
  const user = {
    login: "testuser",
    id: 1,
    avatar_url: "https://via.placeholder.com/150",
    node_id: "1",
    gravatar_id: "11",
    url: "https://via.placeholder.com/150",
    html_url: "wdq",
    followers_url: "qwd",
    following_url: "wdq",
    gists_url: "wqd",
    starred_url: "qwd",
    subscriptions_url: "dqw",
    organizations_url: "wd",
    repos_url: "wdq",
    events_url: "wdq",
    received_events_url: "wdq",
    type: "wdq",
    site_admin: false,
    score: 0
  };

  it('should render correctly with light theme by default', () => {
    const { asFragment, getByText, getByAltText } = renderWithRedux(<Card user={user} />);
    expect(getByText('testuser')).toBeInTheDocument();
    expect(getByAltText('user')).toBeInTheDocument();
    expect(getByAltText('user')).toHaveAttribute('src', 'https://via.placeholder.com/150');
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with dark theme when enabled', () => {
    const darkThemeState = {
      ...initialState,
      themeSlice: {
        isDark: true,
      },
    };
    const { asFragment, container } = renderWithRedux(<Card user={user} />, { store: configureStore({ reducer: rootReducer, preloadedState: darkThemeState }) });
    expect(container.firstChild).toHaveClass('user-dark');
    expect(asFragment()).toMatchSnapshot();
  });
});


/*
import React from 'react';
import { render,  

} from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { rootReducer, RootState } from '../../redux/store';
import Card from './index';
 

const initialState: RootState = {
  appSlice: {
    isLoading: false,
    error: null,
    isLoadingUserData: false,
    users: [],
    typedValue: "",
    params: {
      limit: 10,
      offset: 1,
      query: '',
      storedValue: '',
    },
    status: 'idle',
    clickedUser: {
      login: "",
      id: 0,
      node_id: "",
      avatar_url: "",
      gravatar_id: "",
      url: "",
      html_url: "",
      followers_url: "",
      following_url: "",
      gists_url: "",
      starred_url: "",
      subscriptions_url: "",
      organizations_url: "",
      repos_url: "",
      events_url: "",
      received_events_url: "",
      type: "",
      site_admin: false,
      score: 0
    }
  },
  selectedElementsSlice: {
    storedElements: [],
    selectedElement: null,
  },
  themeSlice: {
    isDark: false, 
  },
};

const renderWithRedux = (
  component: React.ReactNode,
  { store = configureStore({ reducer: rootReducer, preloadedState: initialState }) } = {}
) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
};

describe('Card component', () => {
  const user = {
    login: "testuser",
    id: 1,
    avatar_url: "https://via.placeholder.com/150",
    
 
    node_id: "1",
    gravatar_id: "11",
    url: "https://via.placeholder.com/150",
    html_url: "wdq",
    followers_url: "qwd",
    following_url: "wdq",
    gists_url: "wqd",
    starred_url: "qwd",
    subscriptions_url: "dqw",
    organizations_url: "wd",
    repos_url: "wdq",
    events_url: "wdq",
    received_events_url: "wdq",
    type: "wdq",
    site_admin: false,
    score: 0
 
  };

  it('should render correctly with light theme by default', () => {
    const { getByText, getByAltText } = renderWithRedux(<Card user={user} />);
    expect(getByText('testuser')).toBeInTheDocument();
    expect(getByAltText('user')).toBeInTheDocument();
    expect(getByAltText('user')).toHaveAttribute('src', 'https://via.placeholder.com/150');
  });
  it('should render correctly with dark theme when enabled', () => {
      const darkThemeState = {
      ...initialState,
      themeSlice: {
        isDark: true,
      },
    };
    const { container } = renderWithRedux(<Card user={user} />, { store: configureStore({ reducer: rootReducer, preloadedState: darkThemeState }) });
    expect(container.firstChild).not.toHaveClass('user-dark');
  }); 
});
*/