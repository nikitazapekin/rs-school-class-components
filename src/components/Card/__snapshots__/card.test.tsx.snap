// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should render correctly with light theme by default 1`] = `
<DocumentFragment>
  <div
    class="user__link"
  >
    <div
      class="user "
    >
      <img
        alt="user"
        class="user__logo"
        data-nimg="1"
        decoding="async"
        height="300"
        loading="lazy"
        src="https://via.placeholder.com/150"
        srcset="https://via.placeholder.com/150 1x, https://via.placeholder.com/150 2x"
        style="color: transparent;"
        width="300"
      />
      <div
        class="user__content"
      >
        <p
          class="user__login"
        >
          testuser
        </p>
        <input
          class="user__checkbox"
          readonly=""
          type="checkbox"
        />
        <a
          class="user__details"
          href="/details?page=1&user=testuser"
          style="background-color: rgb(90, 90, 190); padding: 10px; color: rgb(255, 255, 255); text-decoration: none; border-radius: 10px;"
        >
          Show Details
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
